pkg/controller/chi/worker-reconciler.go

```diff
@@ -377,6 +379,81 @@ func (w *worker) reconcileCluster(ctx context.Context, cluster *chiV1.Cluster) e
 	return nil
 }
 
+func (w *worker) safeShardWorkerCount(shards []*chiV1.ChiShard) int {
+	maxAvailableWorkers := float64(chop.Config().Reconcile.Runtime.ShardsThreadNumber)
+	maxPct := float64(chop.Config().Reconcile.Runtime.ShardsMaxConcurrencyPercent)
+
+	// Always return at least 1, up to maxAvailable, but never exceeding maxSafe
+	maxSafeWorkers := math.Max(math.Floor(maxPct/100.0)*float64(len(shards)), 1)
+	return int(math.Min(maxAvailableWorkers, maxSafeWorkers))
+}
+
+func (w *worker) reconcileShardsHosts(ctx context.Context, shards []*chiV1.ChiShard) error {
+	if len(shards) == 0 {
+		return nil
+	}
+
+	// Always process the first shard first (and all of its hosts). This gives us some early indicator of whether
+	// the reconciliation would fail, and for large clusters is small price to pay before performing concurrent fan-out.
+	firstShard := shards[0]
+	if err := w.reconcileShard(ctx, firstShard); err != nil {
+		return err
+	}
+	for _, host := range firstShard.Hosts {
+		if err := w.reconcileHost(ctx, host); err != nil {
+			return err
+		}
+	}
+
+	// If we're working on a single-shard cluster, we are done.
+	if len(shards) == 1 {
+		return nil
+	}
+
+	// Otherwise, we process the remaining n-1 shards using the configured concurrency level while maintaining
+	// the configured max concurrency percentage.
+	// The default behavior is no concurrency, but this can be adjusted in operating reconciliation runtime settings.
+	workerCount := w.safeShardWorkerCount(shards)
```

Comment:
Note that with the default configuration, this is inert in that it reconciles each shard sequentially (granted, I recognize that the goroutine launching inside the loop is net-new, even when using the safe/inert default of concurrency level = 1). I didn't really want to increase the code complexity further to detect that specific case or duplicate code to have a "sequential code path" and a "concurrent code path," which may have its own risks such as somebody making a change in only one of the two hypothetical code paths.